File: Топологическая Сортировка за O(n+m).cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> ans;
void _dfs (vector<bool> &used, vector<vector<int> > &g,int v)
{
	used[v] = true;
	for (int i = 0; i<g[v].size(); ++i)
    {
		int to = g[v][i];
		if (!used[to])
			_dfs (used,g,to);
	}
	ans.push_back (v);
}

void topological_sort(vector<bool> &used,vector<vector<int> > &g, int n)
{
	for (int i = 0; i<n; ++i)
		used[i] = false;
	ans.clear();
	for (int i = 0; i<n; ++i)
		if (!used[i])
			_dfs (used,g,i);
	reverse (ans.begin(), ans.end());
}


/*
-g - массив со списками смежностей.
 n - число вершин в графе.
 used - булевский массив забитый false-ами.
 ans - пустой массив для ответа.
-Граф хранить в виде списков смежностей.
 Если в i-ом массиве есть элемент j, значит из i можно попасть в j.
-Функция top перенумеровывает вершины таким образом, чтобы каждое ребро вело из вершины с меньшим номером в вершину с большим.
*/

File: Топологическая Сортировка за O(n+m).py
ans = []
def _dfs(used:list, g:list, v:int) -> None:
    used[v] = True
    for i in range(len(g[v])):
        to = g[v][i]
        if (not used[to]):
            _dfs(used,g,to)
    ans.append(v)
    
def top(used:list, g:list, n:int) -> None:
    for i in range(n):
        used[i] = False
    ans.clear()
    for i in range(n):
        if (not used[i]):
            _dfs(used,g,i)
    ans.reverse()


'''
-g - массив со списками смежностей.
 n - число вершин в графе.
 used - булевский массив забитый False-ами.
 ans - пустой массив для ответа.
-Граф хранить в виде списков смежностей.
 Если в i-ом массиве есть элемент j, значит из i можно попасть в j.
-Функция top перенумеровывает вершины таким образом, чтобы каждое ребро вело из вершины с меньшим номером в вершину с большим.
'''